internal/era/era.go

```diff
@@ -0,0 +1,343 @@
+	TypeAccumulator        uint16 = 0x07
+	TypeBlockIndex         uint16 = 0x3266
+
+	MaxEra1Size = 8192
+)
+
+// Filename returns a recognizable Era1-formatted file name for the specified
+// epoch and network.
+func Filename(network string, epoch int, root common.Hash) string {
+	return fmt.Sprintf("%s-%05d-%s.era1", network, epoch, root.Hex()[2:10])
+}
+
+// ReadDir reads all the era1 files in a directory for a given network.
+// Format: <network>-<epoch>-<hexroot>.era1
+func ReadDir(dir, network string) ([]string, error) {
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, fmt.Errorf("error reading directory %s: %w", dir, err)
+	}
+	var (
+		next = uint64(0)
+		eras []string
+	)
+	for _, entry := range entries {
+		if path.Ext(entry.Name()) != ".era1" {
+			continue
+		}
+		parts := strings.Split(entry.Name(), "-")
+		if len(parts) != 3 || parts[0] != network {
+			// invalid era1 filename, skip
+			continue
+		}
+		if epoch, err := strconv.ParseUint(parts[1], 10, 64); err != nil {
+			return nil, fmt.Errorf("malformed era1 filename: %s", entry.Name())
+		} else if epoch != next {
+			return nil, fmt.Errorf("missing epoch %d", next)
+		}
+		next += 1
+		eras = append(eras, entry.Name())
+	}
+	return eras, nil
+}
+
+// Reader reads an Era1 archive.
+// See Builder documentation for a detailed explanation of the Era1 format.
+type Reader struct {
+	r io.ReaderAt
+	e *e2store.Reader
+
+	buf      [8]byte  // buffer reading entry offsets
+	next     uint64   // next block to read
+	length   int64    // total length of r
+	metadata metadata // start, count info
+}
+
+type ReadAtSeeker interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// NewReader returns a new Reader instance.
+func NewReader(r ReadAtSeeker) (*Reader, error) {
+	length, err := r.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	m, err := readMetadata(r, length)
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{
+		r:        r,
+		e:        e2store.NewReader(r),
+		next:     m.start,
+		length:   length,
+		metadata: m,
+	}, nil
+}
+
+// readOffset reads a specific block's offset from the block index. The value n
+// is the absolute block number desired.
+func (r *Reader) readOffset(n uint64) (int64, error) {
+	var (
+		firstIndex  = -8 - int64(r.metadata.count)*8      // size of count - index entries
+		indexOffset = int64(n-r.metadata.start) * 8       // desired index * size of indexes
+		offOffset   = r.length + firstIndex + indexOffset // offset of block offset
+	)
+	r.clearBuffer()
+	if _, err := r.r.ReadAt(r.buf[:], offOffset); err != nil {
+		return 0, err
+	}
+	// Since the block offset is relative from its location + size of index
+	// value (8), we need to add it to it's offset to get the block's
+	// absolute offset.
+	return offOffset + 8 + int64(binary.LittleEndian.Uint64(r.buf[:])), nil
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
```

Comment:
I think this method should be called `Next()`, since it's kind of like an iterator, has an internal state which progresses every time you call it

Reply:
done in 3551a7c8f7ca08c0bac18bf618434388dd205500